### PR TITLE
Browser-level tests fixture

### LINKED ISSUE #278: create Browser-level tests fixture
playwright runs most of it's tests on the same browser instance without starting a new process for each test.

we should add `PlaywrightSharpBrowserLoaderCollection` or something similar

### PR DESCRIPTION
**Pull request by Meir017**
closes #278

### CODE REVIEW

**Comment by Meir017** [comment]
@kblok some tests in `BrowserContextTests` are failing due to too many browser context existing
I wonder how playwright solved this. are they not running in parallel?

**Comment by kblok** [comment]
Tests don't run in parallel in pw @Meir017 

If keeping one browser opened is something we can do, good, if not I don't think that's something I would worry much.
That was a desired effect back in the past, trying to monitory process leaks.

**Comment by Meir017** [comment]
I think is starting to get important. our firefox CI is taking longer than firefox on playwright and we still have 500+ tests

**Comment by Meir017** [comment]
@kblok check this out :D

**Comment by Meir017** [comment]
Firefox on MacOS took ~2m instead of 5m+

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpBrowserBaseTest.cs [hunk 1]
```diff
@@ -12,22 +10,23 @@ namespace PlaywrightSharp.Tests.BaseTests
...
+        internal IBrowser Browser => PlaywrightSharpBrowserLoaderFixture.Browser;
+        internal IBrowserApp BrowserApp => PlaywrightSharpBrowserLoaderFixture.BrowserApp;
+
+        private readonly List<IBrowserContext> _contexts = new List<IBrowserContext>();
 
         internal LaunchOptions DefaultOptions { get; set; }
 
         internal PlaywrightSharpBrowserBaseTest(ITestOutputHelper output) : base(output)
         {
-            BaseDirectory = Path.Combine(Directory.GetCurrentDirectory(), "workspace");
```
Comment (Meir017) [inline src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpBrowserBaseTest.cs:21]: `PlaywrightSharpBaseTest` does this already

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoaderFixture.cs [hunk 2]
```diff
@@ -7,27 +7,17 @@ namespace PlaywrightSharp.Tests.BaseTests
...
-        {
-            SetupAsync().GetAwaiter().GetResult();
-        }
-
-        /// <inheritdoc />
-        public void Dispose()
-        {
-            Task.WaitAll(Server.StopAsync(), HttpsServer.StopAsync());
-        }
+        static PlaywrightSharpLoaderFixture() => SetupAsync().GetAwaiter().GetResult();
```
Comment (Meir017) [inline src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoaderFixture.cs:16]: had issues where multiple initializers were calling this and throwing due to the server running already.

**Comment by Meir017** [comment]
see `test\playwright.spec.js` in playwright for details on which specs require a browser/page and which ones start from scratch

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightXunitTestFramework.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    /// <inheritdoc/>
+    public class PlaywrightXunitTestFramework : XunitTestFramework
+    {
+        /// <inheritdoc/>
+        public PlaywrightXunitTestFramework(IMessageSink messageSink) : base(messageSink)
+        {
+        }
+
+        protected override ITestFrameworkExecutor CreateExecutor(AssemblyName assemblyName)
+            => new PlaywrightXunitTestFrameworkExecutor(assemblyName, SourceInformationProvider, DiagnosticMessageSink);
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/BaseTests/PlaywrightXunitTestFramework.cs:21]: Let's make GH Actions happy.

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpBrowserBaseTest.cs [hunk 4]
```diff
@@ -36,10 +35,22 @@ internal async Task<IPage> NewPageAsync(BrowserContextOptions options = null)
...
         /// <inheritdoc cref="IAsyncLifetime.DisposeAsync"/>
-        public virtual Task DisposeAsync() => Browser.CloseAsync();
+        public async Task DisposeAsync()
+        {
+            await Task.WhenAll(_contexts.ConvertAll(context => context.CloseAsync()));
+            await AdditionalDisposeAsync();
+        }
+
+        /// <inheritdoc/>
+        protected virtual Task AdditionalInitializeAsync() => Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (Meir017) [reply src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpBrowserBaseTest.cs:51]: it was a work in progress, I'm removing it now

**Comment by Meir017** [comment]
setting up .NetCore on linux failed 😕

**Pull request merged**